hal/src/thumbv6m/sercom/v2/spi.rs

```diff
@@ -769,56 +773,73 @@ where
 ///
 /// A set of [`Pads`] cannot be used to transmit when the Data Out [`Pad`] is
 /// [`NoneT`].
-pub trait NotTx: AnyPads {}
+pub trait NotTx: ValidPads {}
 
-impl<P> NotTx for P where P: AnyPads<DataOut = NoneT> {}
+impl<P> NotTx for P where P: ValidPads<DataOut = NoneT> {}
 
 /// Marker trait for a set of [`Pads`] that cannot receive
 ///
 /// A set of [`Pads`] cannot be used to receive when the Data In [`Pad`] is
 /// [`NoneT`].
-pub trait NotRx: AnyPads {}
+pub trait NotRx: ValidPads {}
 
-impl<P> NotRx for P where P: AnyPads<DataIn = NoneT> {}
+impl<P> NotRx for P where P: ValidPads<DataIn = NoneT> {}
 
 /// Marker trait for a set of [`Pads`] that can transmit OR receive
 ///
 /// To satisfy this trait, SCLK must always be [`SomePad`] and one or both of
 /// Data In and Data Out must also be [`SomePad`].
-pub trait TxOrRx: AnyPads {}
+pub trait TxOrRx: ValidPads {}
 
 impl<S, DI, CK, SS> TxOrRx for Pads<S, DI, NoneT, CK, SS>
 where
     S: Sercom,
-    DI: SomePad,
-    CK: SomePad,
-    SS: OptionalPad,
+    DI: GetPad<S> + GetPadMarker,
+    CK: GetPad<S>,
+    SS: GetOptionalPad<S>,
+    Self: DipoDopo,
 {
 }
 
 impl<S, DO, CK, SS> TxOrRx for Pads<S, NoneT, DO, CK, SS>
 where
     S: Sercom,
-    DO: SomePad,
-    CK: SomePad,
-    SS: OptionalPad,
+    DO: GetPad<S> + GetPadMarker,
```

Comment:
I'm not sure why you added the "Get" prefix on these, makes it sound like a function rather than a type.

Reply:
We discussed this on Matrix, but I'll add some detail here, for anyone following along. Your intuition is correct here, because the documentation for `GetPad` says:

```
/// Type-level function to uniquely specify a [`Pad`] type
```

I added documentation on type-level functions and other type-level programming patterns to the module-level docs of the `typelevel` module. I think that should explain it. If not, I can add more detail.